# Notes

## Flight paths:
- **Houston, TX** to **Los Angeles, CA** (IAH - LAX)
- **New York City, NY** to **Miami, FL** (JFK - MIA)
- **Portland, WA** to **Chicago, IL** (PDX - ORD)

## Number of Total Routes
- at least 1,000 per route for now (All times in GMC)

    -__Times:__

        - 0000 hours to 0600 hours

        - 0601 hours to 1200 hours

        - 1201 hours to 1800 hours

        - 1801 hours to 2399 hours

    - __Times of the Year:__
    
        - Try to get every month

    - __Times of the Week:__
        - Try to get every day. 
    
## Notes about Data considerations

- see what are typical flight times for your paths. you may be limited here. 

- consistently work with other api to see what you can grab

- grab future flights too!

- do some division on how many flights you can grab from how many time zonestuff


## Features to Scrape
- want aircraft type
- want airline flight info
- want airline flight 
- want aircraft type struct
- want flight struct

1 query will generate 15 results. Ex; If you request to see all flight from airport Alpha to Airport Bravo and the search results come back with 5000 flights. To find the pricing estimate you would do the following math. 5000/15 = 333 * $0.0079 = $2.63 (Class 2)

In [269]:
import sys
from suds import null, WebFault
from suds.client import Client
import logging
import json
import pandas as pd
import datetime
import numpy as np

In [2]:
with open('/Users/ChristopherKuzemka/Documents/GA/dsi_11/projects/capstone/env.json') as f:
    information = json.load(f)

In [3]:
information.keys()

dict_keys(['FA_API_KEY', 'FA_USERNAME', 'x-rapidapi-host', 'x-rapidapi-key'])

In [4]:
username = information.get('FA_USERNAME')
apiKey = information.get('FA_API_KEY')
url = 'http://flightxml.flightaware.com/soap/FlightXML2/wsdl'

In [5]:
logging.basicConfig(level=logging.INFO)
api = Client(url, username=username, password=apiKey)

## Testing playground determining usefulness of API

In [83]:
#Get Aircraft type  - semi-useful
aircraft_type = api.service.AircraftType('B744')
aircraft_type

(AircraftTypeStruct){
   manufacturer = "Boeing"
   type = "747-400"
   description = "quad-jet"
 }

# JFK - MIA

From [here](https://www.flights.com/flights/new-york-jfk-to-miami-mia/): "with 3 differnt airlines operating flights between New York and Miami, there are, on average, 2,197 flights per month.. This equates to about 523 flights per week, and 75 flights per day from JFK to MIA. The three airlines are:
- American Airlines (Flight AA 2572)

- British Airways (Flight BA 1687) 

- Malaysia Airlines (Flight MH 9446).

In [7]:
def make_unix_lists(start_date, end_date, frequency):
    created_range = pd.date_range(start = start_date, end = end_date, freq = frequency) #creates a daterange series
    list_created_range = list(created_range) #converts such range into a list
    unix_floats = [date.to_pydatetime().timestamp() for date in list_created_range] #transforms the daterange list into unix epoch tiimestamps represented as floats
    unix_ints = [int(i) for i in unix_floats] #makes the above list as a list of integers
    #return unix_ints

    #We are doing the below to accomodate a for loop format into another function
    start_ints = unix_ints[:-1] #creates a list of all the start dates without last element
    end_ints = unix_ints[1:] #creates a list of all the end dates without first element 

    return start_ints, end_ints

In [8]:
start, end = make_unix_lists('5/11/2020', '5/24/2020', '8H')

In [9]:
start

[1589169600,
 1589198400,
 1589227200,
 1589256000,
 1589284800,
 1589313600,
 1589342400,
 1589371200,
 1589400000,
 1589428800,
 1589457600,
 1589486400,
 1589515200,
 1589544000,
 1589572800,
 1589601600,
 1589630400,
 1589659200,
 1589688000,
 1589716800,
 1589745600,
 1589774400,
 1589803200,
 1589832000,
 1589860800,
 1589889600,
 1589918400,
 1589947200,
 1589976000,
 1590004800,
 1590033600,
 1590062400,
 1590091200,
 1590120000,
 1590148800,
 1590177600,
 1590206400,
 1590235200,
 1590264000]

In [10]:
jfk_mia_info = {'AA':'2572',
                'BA':'1687',
                'MA':'9446'}

In [380]:
def get_flights_df(start_input, end_input, frequency_input, origin_input, destination_input, flight_info_input, howMany_input):
    ## Empty lists for storage reasons
    schedules = [] #instantiates the schedules column
    modified_schedules = [] #instantiates the modified schedules list
    flight_info_list = [] #instantiates the flight info list


    ## Creates the sketelton dataframe for schedules_df
    schedules_cols = ['ident', 'actual_ident', 'departuretime', 'arrival_time', 'origin', 'destination', 'aircrafttype', 'meal_service', 'seats_cabin_first', 'seats_cabin_business', 'seats_cabin_coach'] #column names for the scheduling df
    schedules_df = pd.DataFrame(columns = column_names) #creates the empty schedules dataframe skeleton

    ## Creates the skeleton dataframe for the flight_info_df
    flight_info_cols = ['faFlightID', 'ident', 'aircrafttype', 'filed_ete', 'filed_time', 'filed_departuretime', 'filed_airspeed_mach', 'filed_altitude', 'actualdeparturetime', 'estimatedarrivaltime', 'actualarrivaltime', 'diverted', 'origin', 'originName', 'originCity', 'destinationName', 'destinationCity']
    flight_info_df = pd.DataFrame(columns = names) #creates the empty flight_info dataframe skeleton

    ## Unix Epoch Time Function
    start_list, end_list = make_unix_lists(start_input, end_input, frequency_input) #utilizes the unix epoch creation list function

    ## Magic happens here
    for airline_code in flight_info_input: #for the keys in the flight info input 
        for j in range(len(start_list)): #for the entirity of the unix epoch time list
            airline_flight_schedules = api.service.AirlineFlightSchedules(startDate = start_list[j], endDate = end_list[j], origin = origin_input, destination = destination_input, airline = airline_code, flightno = flight_info_input.get(airline_code), howMany = howMany_input) #gets the flight schedules information

            #Credit for the dict below
            #https://stackoverflow.com/questions/17581731/parsing-suds-soap-complex-data-type-into-python-dict
            airline_flight_dict = Client.dict(airline_flight_schedules) #converts the above variable to a dictionary
            schedules.append(airline_flight_dict) #appends results to the instantiated output list
    
    ## Cleans the dirty collection for schedules
    for k in range(len(schedules)):
        try: #try the below...we'll get some errors for empty stuff
            modified_schedules.append(schedules[k]['data']) #filters the output list to only show meaningful information where actual flights were flows in a timeframes studied
        except: #when an error occurs...
            continue #..move on and ignore any issues with with non existent flight data

    ## Creates a dataframe out of the modified_schedules list <-- FIX THIS LATER
    for l in range(len(modified_schedules)):
        for m in range(len(modified_schedules[l])):
            sched_df = pd.DataFrame(modified_schedules[l][m]).T #uses a transposed version to cleanly extract the data
            sched_df.rename(columns = {0:schedules_cols[0], #renames the columns to match the end_df so concatenation is simple
                        1: schedules_cols[1],
                        2: schedules_cols[2],
                        3: schedules_cols[3],
                        4: schedules_cols[4],
                        5: schedules_cols[5],
                        6: schedules_cols[6],
                        7: schedules_cols[7],
                        8: schedules_cols[8],
                        9: schedules_cols[9],
                        10: schedules_cols[10]}, inplace = True) #very inefficient way to rename my columns
            schedules_df = pd.concat([schedules_df, sched_df]) #overwrites the end_df to create a larger dataframe
    
    ## Cleans the schedules_df
    schedules_df.drop(index = 0, inplace = True) #drops the 0 index rows in the repetative 0,1 pattern of the dataframe
    schedules_df.reset_index(inplace = True) #resets the index by creating a new index row
    schedules_df.drop(columns = 'index', inplace = True) #drops the extra 'index row
    schedules_df.drop_duplicates(subset = ['arrival_time'], inplace = True) #drops the duplicate time epochs so that we don't have impossible situations of the same flight coming into the same place at the same exact time


    ## Gets the flight information list using the schedules_df idents
    for identifier in schedules_df['ident']: #calls to each identifier collected
        for offset_number in range(15): #set in place to not miss any flights in 2 week period 
            try: #we're gonna get some errors here for empty flights I think
                flight_info_ex = api.service.FlightInfoEx(ident = identifier, howMany = 15, offset = offset_number) #retrieves the information pertaining to the identifier
                flight_info = Client.dict(flight_info_ex) #converts the suds object into a workable dictionary
                flight_info_list.append(flight_info_ex) #appends whatever the above is into a list
            except: #when an error occurs...
                continue #...just move on

    ## Creates a dataframe out of the flight_info_df
    ## Credit to Alex Speicher from Hofstra University for showing an efficient way to do this
    for query in flight_info_list: #calls to every query in the info list made
        for flight in query['flights']: #for every flight in the dictionary
            flight_dict = {} #empty dicitonary pertaining to the columns in each flight 
            for column in flight_info_cols: #for every column name in the flight_info_cols
                flight_dict[column] = flight[column] #appends the column names to the dictionary
            info_df = pd.DataFrame([flight_dict], columns=flight_dict.keys()) #makes a dataframe using the dictionary columns 
            flight_info_df = pd.concat([flight_info_df, info_df], axis =0).reset_index(drop = True) #resets the index and fixes all those issues

    #Cleans the flight_info_df
    flight_info_df.drop_duplicates(subset = ['faFlightID'], inplace = True) #eliminates the overlapping most unique identifiers in our
    flight_info_df = flight_info_df[flight_info_df['actualarrivaltime'] != 0] #removes the currently active flights from the list as there is no data on them and overwrites the dataframe variable
    flight_info_df.reset_index(inplace = True) #resets the index 

    return schedules_df, flight_info_df #returns two dataframes

In [381]:
jfk_mia_flight_scheds, jfk_mia_flight_info = get_flights_df('5/15/2020', '5/25/2020', '8H', 'JFK', 'MIA', jfk_mia_info, 15)

ERROR:suds.client:<suds.sax.document.Document object at 0x12099a110>
ERROR:suds.client:<suds.sax.document.Document object at 0x117c7c710>
ERROR:suds.client:<suds.sax.document.Document object at 0x1202f7510>
ERROR:suds.client:<suds.sax.document.Document object at 0x120b27150>
ERROR:suds.client:<suds.sax.document.Document object at 0x122c703d0>
ERROR:suds.client:<suds.sax.document.Document object at 0x1070bdd90>
ERROR:suds.client:<suds.sax.document.Document object at 0x1221c2750>
ERROR:suds.client:<suds.sax.document.Document object at 0x122291710>
ERROR:suds.client:<suds.sax.document.Document object at 0x12212f810>
ERROR:suds.client:<suds.sax.document.Document object at 0x124108650>
ERROR:suds.client:<suds.sax.document.Document object at 0x125947610>
ERROR:suds.client:<suds.sax.document.Document object at 0x12626c190>
ERROR:suds.client:<suds.sax.document.Document object at 0x125267650>
ERROR:suds.client:<suds.sax.document.Document object at 0x12533cad0>
ERROR:suds.client:<suds.sax.docume

In [514]:
jfk_mia_flight_scheds

ident actual_ident departuretime arrival_time origin destination  \
0   MAS9446      AAL2572    1589567400   1589579280   KJFK        KMIA   
1   GLO6391      AAL2572    1589569200   1589581200   KJFK        KMIA   
3   MAS9446      AAL2572    1589653800   1589665680   KJFK        KMIA   
4   GLO6391      AAL2572    1589655600   1589667600   KJFK        KMIA   
6   MAS9446      AAL2572    1589740200   1589752080   KJFK        KMIA   
7   GLO6391      AAL2572    1589742000   1589754000   KJFK        KMIA   
9   MAS9446      AAL2572    1589826600   1589838480   KJFK        KMIA   
10  GLO6391      AAL2572    1589828400   1589840400   KJFK        KMIA   
12  MAS9446      AAL2572    1589913000   1589924880   KJFK        KMIA   
13  AAL2572         None    1589914800   1589926800   KJFK        KMIA   
14  MAS9446      AAL2572    1589999400   1590011280   KJFK        KMIA   
15  GLO6391      AAL2572    1590001200   1590013200   KJFK        KMIA   
17  MAS9446      AAL2572    1590085800   1590097680   KJFK        KMIA   
18  GLO6391      AAL2572    1590087600   1590099600   KJFK        KMIA   
20  MAS9446      AAL2572    1590172200   1590184080   KJFK        KMIA   
21  GLO6391      AAL2572    1590174000   1590186000   KJFK        KMIA   
23  MAS9446      AAL2572    1590258600   1590270480   KJFK        KMIA   
24  GLO6391      AAL2572    1590260400   1590272400   KJFK        KMIA   
26  MAS9446      AAL2572    1590345000   1590356880   KJFK        KMIA   
27  GLO6391      AAL2572    1590346800   1590358800   KJFK        KMIA   

   aircrafttype                          meal_service seats_cabin_first  \
0          B738        Business: Meal / Economy: Meal                 0   
1          B738  Business: No meal / Economy: No meal                 0   
3          B738        Business: Meal / Economy: Meal                 0   
4          B738  Business: No meal / Economy: No meal                 0   
6          B738        Business: Meal / Economy: Meal                 0   
7          B738  Business: No meal / Economy: No meal                 0   
9          B738        Business: Meal / Economy: Meal                 0   
10         B738  Business: No meal / Economy: No meal                 0   
12         B738        Business: Meal / Economy: Meal                 0   
13         B738  Business: No meal / Economy: No meal                 0   
14         B738        Business: Meal / Economy: Meal                 0   
15         B738  Business: No meal / Economy: No meal                 0   
17         B738        Business: Meal / Economy: Meal                 0   
18         B738  Business: No meal / Economy: No meal                 0   
20         B738        Business: Meal / Economy: Meal                 0   
21         B738  Business: No meal / Economy: No meal                 0   
23         B738        Business: Meal / Economy: Meal                 0   
24         B738  Business: No meal / Economy: No meal                 0   
26         B738        Business: Meal / Economy: Meal                 0   
27         B738  Business: No meal / Economy: No meal                 0   

   seats_cabin_business seats_cabin_coach  
0                    16               144  
1                    16               144  
3                    16               144  
4                    16               144  
6                    16               144  
7                    16               144  
9                    16               144  
10                   16               144  
12                   16               144  
13                   16               144  
14                   16               144  
15                   16               144  
17                   16               144  
18                   16               144  
20                   16               144  
21                   16               144  
23                   16               144  
24                   16               144  
26                   16               1

In [366]:
jfk_mia_flight_info.drop_duplicates(subset = ['faFlightID'], inplace = True)

In [369]:
mask = jfk_mia_flight_info['actualarrivaltime'] != 0
jfk_mia_flight_info[mask]

faFlightID    ident aircrafttype filed_ete  \
3   AAL2572-1590209159-airline-0537  AAL2572         B738  03:06:00   
4   AAL2572-1590122721-airline-0420  AAL2572         B738  02:40:00   
5   AAL2572-1590036344-airline-0175  AAL2572         B738  02:55:00   
6   AAL2572-1589949964-airline-0120  AAL2572         B738  02:40:00   
7   AAL2572-1589863519-airline-0459  AAL2572         B738  02:40:00   
8   AAL2572-1589777123-airline-0485  AAL2572         B738  02:38:00   
9   AAL2572-1589690730-airline-0195  AAL2572         B738  02:46:00   
10  AAL2572-1589604351-airline-0000  AAL2572         B738  02:41:00   
11  AAL2572-1589517913-airline-0098  AAL2572         B738  02:40:00   
12  AAL2572-1589431536-airline-0436  AAL2572         B738  02:40:00   
13  AAL2572-1589345159-airline-0102  AAL2572         B738  02:56:00   

    filed_time filed_departuretime filed_airspeed_mach filed_altitude  \
3   1590431725          1590432000                None            360   
4   1590345342          1590345600                None            380   
5   1590259235          1590259200                None            380   
6   1590172660          1590172800                None            380   
7   1590086190          1590086400                None            380   
8   1590000375          1590000000                None            380   
9   1589914315          1589913600                None            380   
10  1589827142          1589827200                None            320   
11  1589741446          1589740800                None            380   
12  1589654365          1589654400                None            300   
13  1589567883          1589568000                None            380   

   actualdeparturetime estimatedarrivaltime actualarrivaltime diverted origin  \
3           1590431773           1590440887        1590440887     None   KJFK   
4           1590345390           1590354360        1590354360     None   KJFK   
5           1590259283           1590267480        1590267480     None   KJFK   
6           1590172708           1590181620        1590181620     None   KJFK   
7           1590086238           1590095340        1590095340     None   KJFK   
8           1590000407           1590009900        1590009900     None   KJFK   
9           1589914363           1589924160        1589924160     None   KJFK   
10          1589827191           1589837040        1589837040     None   KJFK   
11          1589741494           1589750762        1589750762     None   KJFK   
12          1589654430           1589663460        1589663460     None   KJFK   
13          1589567931           1589576820        1589576820     None   KJFK   

             originName    originCity destinationName destinationCity  
3   John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
4   John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
5   John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
6   John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
7   John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
8   John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
9   John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
10  John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
11  John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
12  John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
13  John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL

In [378]:
mask = jfk_mia_flight_scheds.duplicated(subset = ['arrival_time']) == True
jfk_mia_flight_scheds[mask]

ident actual_ident departuretime arrival_time origin destination  \
2   AAL2572         None    1589569200   1589581200   KJFK        KMIA   
5   AAL2572         None    1589655600   1589667600   KJFK        KMIA   
8   AAL2572         None    1589742000   1589754000   KJFK        KMIA   
11  AAL2572         None    1589828400   1589840400   KJFK        KMIA   
16  AAL2572         None    1590001200   1590013200   KJFK        KMIA   
19  AAL2572         None    1590087600   1590099600   KJFK        KMIA   
22  AAL2572         None    1590174000   1590186000   KJFK        KMIA   
25  AAL2572         None    1590260400   1590272400   KJFK        KMIA   
28  AAL2572         None    1590346800   1590358800   KJFK        KMIA   

   aircrafttype                          meal_service seats_cabin_first  \
2          B738  Business: No meal / Economy: No meal                 0   
5          B738  Business: No meal / Economy: No meal                 0   
8          B738  Business: No meal / Economy: No meal                 0   
11         B738  Business: No meal / Economy: No meal                 0   
16         B738  Business: No meal / Economy: No meal                 0   
19         B738  Business: No meal / Economy: No meal                 0   
22         B738  Business: No meal / Economy: No meal                 0   
25         B738  Business: No meal / Economy: No meal                 0   
28         B738  Business: No meal / Economy: No meal                 0   

   seats_cabin_business seats_cabin_coach  
2                    16               144  
5                    16               144  
8                    16               144  
11                   16               144  
16                   16               144  
19                   16               144  
22                   16               144  
25                   16               144  
28                   16               144

In [417]:
jfk_mia_flight_info

faFlightID    ident aircrafttype filed_ete  \
3   AAL2572-1590209159-airline-0537  AAL2572         B738  03:06:00   
4   AAL2572-1590122721-airline-0420  AAL2572         B738  02:40:00   
5   AAL2572-1590036344-airline-0175  AAL2572         B738  02:55:00   
6   AAL2572-1589949964-airline-0120  AAL2572         B738  02:40:00   
7   AAL2572-1589863519-airline-0459  AAL2572         B738  02:40:00   
8   AAL2572-1589777123-airline-0485  AAL2572         B738  02:38:00   
9   AAL2572-1589690730-airline-0195  AAL2572         B738  02:46:00   
10  AAL2572-1589604351-airline-0000  AAL2572         B738  02:41:00   
11  AAL2572-1589517913-airline-0098  AAL2572         B738  02:40:00   
12  AAL2572-1589431536-airline-0436  AAL2572         B738  02:40:00   
13  AAL2572-1589345159-airline-0102  AAL2572         B738  02:56:00   

    filed_time filed_departuretime filed_airspeed_mach filed_altitude  \
3   1590431725          1590432000                None            360   
4   1590345342          1590345600                None            380   
5   1590259235          1590259200                None            380   
6   1590172660          1590172800                None            380   
7   1590086190          1590086400                None            380   
8   1590000375          1590000000                None            380   
9   1589914315          1589913600                None            380   
10  1589827142          1589827200                None            320   
11  1589741446          1589740800                None            380   
12  1589654365          1589654400                None            300   
13  1589567883          1589568000                None            380   

   actualdeparturetime estimatedarrivaltime actualarrivaltime diverted origin  \
3           1590431773           1590440887        1590440887     None   KJFK   
4           1590345390           1590354360        1590354360     None   KJFK   
5           1590259283           1590267480        1590267480     None   KJFK   
6           1590172708           1590181620        1590181620     None   KJFK   
7           1590086238           1590095340        1590095340     None   KJFK   
8           1590000407           1590009900        1590009900     None   KJFK   
9           1589914363           1589924160        1589924160     None   KJFK   
10          1589827191           1589837040        1589837040     None   KJFK   
11          1589741494           1589750762        1589750762     None   KJFK   
12          1589654430           1589663460        1589663460     None   KJFK   
13          1589567931           1589576820        1589576820     None   KJFK   

             originName    originCity destinationName destinationCity  
3   John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
4   John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
5   John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
6   John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
7   John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
8   John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
9   John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
10  John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
11  John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
12  John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL  
13  John F Kennedy Intl  New York, NY      Miami Intl       Miami, FL

In [568]:
def get_flights(start_input, end_input, frequency_input, destination_input):
    ## Unix Epoch Time Function
    start_list, end_list = make_unix_lists(start_input, end_input, frequency_input) #utilizes the unix epoch creation list function
    
    ## Create an empty list to add searches to
    get_flights_list = []    


    ## Skeleton df for get_flights_df
    get_flights_cols = ['faFlightID', 'ident', 'prefix', 'type', 'suffix', 'origin', 'destination', 'timeout', 'timestamp', 'departureTime', 'firstPositionTime', 'arrivalTime', 'longitude', 'latitude', 'lowLongitude', 'lowLatitude', 'highLongitude', 'highLatitude', 'groundspeed', 'altitude', 'heading', 'altitudeStatus', 'updateType', 'altitudeChange', 'waypoints']
    get_flights_df = pd.DataFrame(columns = get_flights_cols)

    #Grabbing flights from the list of destinations
    for destination in destination_input:
        add_dest = '{= dest ' + destination + '}'
        search_flight = api.service.SearchBirdseyeInFlight(query = add_dest, howMany = 15)
        search_dict = Client.dict(search_flight)
        for retrieved_query in range(len(search_dict['aircraft'])):
            get_flights_list.append(search_dict['aircraft'][retrieved_query])


    #Creates the dataframe and concats to the get_flights_df
    for list_item in range(len(get_flights_list)):
        preliminary_flight_df = pd.DataFrame(get_flights_list[list_item]).T #uses a transposed version to cleanly extract the data
        preliminary_flight_df.rename(columns = {0:get_flights_cols[0], #renames the columns to match the end_df so concatenation is simple
                            1: get_flights_cols[1],
                            2: get_flights_cols[2],
                            3: get_flights_cols[3],
                            4: get_flights_cols[4],
                            5: get_flights_cols[5],
                            6: get_flights_cols[6],
                            7: get_flights_cols[7],
                            8: get_flights_cols[8],
                            9: get_flights_cols[9],
                            10: get_flights_cols[10],
                            11: get_flights_cols[11],
                            12: get_flights_cols[12],
                            13: get_flights_cols[13],
                            14: get_flights_cols[14],
                            15: get_flights_cols[15],
                            16: get_flights_cols[16],
                            17: get_flights_cols[17],
                            18: get_flights_cols[18],
                            19: get_flights_cols[19],
                            20: get_flights_cols[20],
                            21: get_flights_cols[21],
                            22: get_flights_cols[22],
                            23: get_flights_cols[23],
                            24: get_flights_cols[24]}, inplace = True) #very inefficient way to rename my columns
        get_flights_df = pd.concat([get_flights_df, preliminary_flight_df]) #overwrites the end_df to create a larger dataframe

    ## Cleans the get_flights_df        
    get_flights_df.drop(index = 0, inplace = True) #drops the 0 index rows in the repetative 0,1 pattern of the dataframe
    get_flights_df.reset_index(inplace = True) #resets the index by creating a new index row
    get_flights_df.drop(columns = 'index', inplace = True) #drops the extra index row

    return get_flights_df

In [573]:
destination_list = ['KLAX', 'KJFK', 'KMIA', 'KIAH', 'KPDX', 'KORD']
test_list = get_flights('5/26/2020','5/27/2020','8H', destination_list)

In [574]:
test_list

faFlightID    ident prefix  type suffix origin  \
0        ACA7229-1590630769-adhoc-0  ACA7229   None  None   None   CYYZ   
1         DAL333-1590379639-fa-0000   DAL333   None  A333   None   KATL   
2          GVPOP-1590589543-adhoc-0   VIR607   None  A35K   None   EGLL   
3         DAL702-1590379578-fa-0000   DAL702   None  A321   None   KATL   
4   CLX7723-1590381916-airline-0199  CLX7723   None  B748   None   EGPK   
..                              ...      ...    ...   ...    ...    ...   
85  SKW2968-1590381915-airline-0154  SKW2968   None  CRJ7   None   CYYZ   
86  DLH8186-1590554734-airline-0156  DLH8186   None  MD11   None   EDDF   
87   BOX484-1590554734-airline-0125   BOX484   None  B772   None   EDDF   
88  ICE853-1590511860-schedule-0056   ICE853   None  B752   None   BIKF   
89       UAL2777-1590438300-fa-0000  UAL2777   None  B789   None   EDDF   

   destination timeout   timestamp departureTime  ... lowLatitude  \
0         KLAX       0  1590631391    1590630660  ...     43.0147   
1         KLAX       0  1590631383    1590625510  ...     33.6341   
2         KLAX       0  1590626335    1590589566  ...     33.9509   
3         KLAX       0  1590626148    1590611945  ...     33.5413   
4         KLAX       0  1590625221    1590580500  ...     33.9367   
..         ...     ...         ...           ...  ...         ...   
85        KORD       0  1590580645    1590576638  ...     42.0026   
86        KORD       0           0             0  ...         200   
87        KORD       0           0             0  ...         200   
88        KORD       0           0             0  ...         200   
89        KORD       0           0             0  ...         200   

   highLongitude highLatitude groundspeed altitude heading altitudeStatus  \
0       -79.6339      43.6546         478      302     243           None   
1       -84.2464      38.2697         481      400     266           None   
2       -2.99184      66.0902         101        1     263           None   
3       -84.2773      35.2755         126        1     263           None   
4       -4.65975      67.0785         135        1     263           None   
..           ...          ...         ...      ...     ...            ...   
85      -79.6209      43.7337         123        7     271           None   
86          -200         -200           0        0       0           None   
87          -200         -200           0        0       0           None   
88          -200         -200           0        0       0           None   
89          -200         -200           0        0       0           None   

   updateType altitudeChange  \
0           A              C   
1           A              -   
2           A              D   
3           A              D   
4           A              D   
..        ...            ...   
85          A              D   
86       None           None   
87       None           None   
88       None           None   
89       None           None   

                                            waypoints  
0   43.68 -79.63 43.65 -79.68 43.64 -79.7 43.63 -7...  
1   33.64 -84.43 33.68 -84.18 33.85 -84.18 34.06 -...  
2   51.48 -0.46 51.51 -0.52 51.55 -0.61 51.6 -0.71...  
3   33.64 -84.43 33.68 -84.28 33.81 -84.28 33.81 -...  
4   55.51 -4.59 55.92 -6.67 56.02 -6.86 56.1 -7.01...  
..                                                ...  
85  43.68 -79.63 43.68 -79.67 43.68 -79.68 43.68 -...  
86                                               None  
87                                               None  
88  63.98 -22.62 63.99 -22.88 63.99 -23.02 63.99 -...  
89  50.03 8.57 50.1 8.42 50.14 8.31 50.2 8.18 50.2...  

[90 rows x 25 columns]

In [ ]:
for airline_code in flight_info_input: #for the keys in the flight info input 
    for j in range(len(start_list)): #for the entirity of the unix epoch time list
        airline_flight_schedules = api.service.AirlineFlightSchedules(startDate = start_list[j], endDate = end_list[j], origin = origin_input, destination = destination_input, airline = airline_code, flightno = flight_info_input.get(airline_code), howMany = howMany_input) #gets the flight schedules information

        #Credit for the dict below
        #https://stackoverflow.com/questions/17581731/parsing-suds-soap-complex-data-type-into-python-dict
        airline_flight_dict = Client.dict(airline_flight_schedules) #converts the above variable to a dictionary
        schedules.append(airline_flight_dict) #appends results to the instantiated output list

# IAH - LAX


From [here](https://www.flights.com/flights/houston-iah-to-los-angeles-lax/): "With 6 different airlines operating between houstan and Los Angeles, there are, on average, 2,642 flights per month. This equates to about 629 flights per week, and 90 flights per day from IAH to LAX." 5 out of 6 of these airliners are: 
- United Airlines (Flight UA 302) 
- Air China (CA 7398), Azul (AD 7026)
- Asiana Airlines (OZ 9512)
- Air New Zealand (NZ 9653)